packages/core/saved-objects/core-saved-objects-server/src/model_version/migration.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { SavedObjectSanitizedDoc } from '../serialization';
+import type { SavedObjectsMigrationLogger } from '../migration';
+
+/**
+ * Document type used during model migration.
+ *
+ * @public
+ */
+export type SavedObjectModelMigrationDoc<T = unknown> = SavedObjectSanitizedDoc<T>;
+
+/**
+ * Context passed down to {@link SavedObjectModelMigrationFn | migration functions}.
+ *
+ * @public
+ */
+export interface SavedObjectModelMigrationContext {
+  /**
+   * logger instance to be used by the migration handler
+   */
+  readonly log: SavedObjectsMigrationLogger;
+  /**
+   * The model version this migration is registered for
+   */
+  readonly modelVersion: number;
+}
+
+/**
+ * Return type for the {@link SavedObjectModelMigrationFn | migration functions}
+ *
+ * @public
+ */
+export interface SavedObjectModelMigrationResult<DocAttrs = unknown> {
+  document: SavedObjectModelMigrationDoc<DocAttrs>;
+}
+
+/**
+ * Migration function for the model version API.
+ *
+ * Similar to the old migration system, model version migrations take the document to migrate
+ * and a context object as input and must return the transformed document in its return value.
+ *
+ * @public
+ */
+export type SavedObjectModelMigrationFn<InputAttributes = unknown, OutputAttributes = unknown> = (
```

Comment:
we use "migration" in a lot of places and it means different things. A "migration" could also mean the whole process of going from one modelVersion to the next (mappings + transforms). What you think about calling these _transformation_ functions like `SavedObjectModelTransformationFn`